akka-actor-tests/src/test/scala/akka/actor/ReceiveTimeoutSpec.scala

```diff
@@ -179,15 +179,59 @@ class ReceiveTimeoutSpec extends AkkaSpec {
         context.setReceiveTimeout(500 milliseconds)
 
         def receive = {
-          case TransperentTick => context.setReceiveTimeout(Duration.Inf)
+          case TransparentTick => context.setReceiveTimeout(Duration.Inf)
           case ReceiveTimeout  => timeoutLatch.open
         }
       }))
 
-      timeoutActor ! TransperentTick
+      timeoutActor ! TransparentTick
 
       intercept[TimeoutException] { Await.ready(timeoutLatch, 1 second) }
       system.stop(timeoutActor)
     }
+
+    "remove / change existing timeout while handling a NotInfluenceReceiveTimeout message" taggedAs TimingTest in {
+      val timeoutLatch = TestLatch()
+      val initialTimeout = 500.millis
+
+      val timeoutActor = system.actorOf(Props(new Actor {
+        context.setReceiveTimeout(initialTimeout)
+
+        def receive: Receive = {
+          case TransparentTick => context.setReceiveTimeout(Duration.Undefined)
+          case ReceiveTimeout  => timeoutLatch.open
+        }
+      }))
+
+      timeoutActor ! TransparentTick
+
+      intercept[TimeoutException] { Await.ready(timeoutLatch, initialTimeout * 2) }
+      system.stop(timeoutActor)
+    }
+
+    "work correctly if the timeout is changed multiple times while handling a NotInfluenceReceiveTimeout message" taggedAs TimingTest in {
+      val timeoutLatch = TestLatch()
+      val initialTimeout = 500.millis
+
+      val timeoutActor = system.actorOf(Props(new Actor {
+        var count = 0
+        context.setReceiveTimeout(initialTimeout)
+
+        def receive: Receive = {
+          case TransparentTick =>
+            context.setReceiveTimeout(initialTimeout * 2)
+            count += 1 // do some work then
+            context.setReceiveTimeout(initialTimeout)
+
+          case ReceiveTimeout => timeoutLatch.open
+        }
+      }))
+
+      timeoutActor ! TransparentTick
+      timeoutActor ! TransparentTick
+
+      Await.ready(timeoutLatch, TestLatch.DefaultTimeout)
```

Comment:
Not quite clear to me how this test would fail? Did it fail before your changes?

Reply:
Are you saying it should fail? It does not fail and did not before.

Reply:
One way of testing it is to use a TestProbe instead of the latch. Send to the probe in `case ReceiveTimeout =>`.

then
```
// not triggered by initialTimeout because it was changed by the ticks
probe.expectNoMessage(initialTimeout + initialTimeout/2)
// now should happen
probe.expectMessage("ReceiveTimeout triggered")
```

Reply:
On second look this addition seems correct - on processing a `NotInfluenceReceiveTimeout`, if the timeout is modified multiple times, and here, given the longer timeout, it would not fail. However if we had it wait for a shorter duration, less than the last it would.

Reply:
This works if it first tests the last value used in `context.setReceiveTimeout` vs `initialTimeout + initialTimeout/2`. The probe receives the `ReceiveTimeout` within ~620 ms.